src/vehicle.cpp

```diff
@@ -2271,6 +2276,98 @@ void Vehicle::HandleLoading(bool mode)
 	this->IncrementImplicitOrderIndex();
 }
 
+/**
+ * Checks whether a vehicle is waiting for automatic separation (if not,
+ * it is ready to depart)
+ */
+bool Vehicle::IsWaitingForAutomaticSeparation() const {
+	Ticks now = _date * DAY_TICKS + _date_fract;
+	return this->AutomaticSeparationIsEnabled() && this->first_order_last_departure > now;
+};
+
+/**
+ * If enabled, calculates the departure time for this vehicle based on the
+ * automatic separation feature.
+ */
+void Vehicle::UpdateAutomaticSeparation()
+{
+	/* Check this feature is enabled on the vehicle's orders */
+	if (!this->AutomaticSeparationIsEnabled()) return;
+
+	/* Only perform the separation at the first manual order (saves on storage) */
```

Comment:
The first manual order is not necessarily a suitable point to wait.

In particular it would be advisable to also exclude waypoint orders, conditional orders and depot orders.

Reply:
I have added to the check the existing `o->CanLoadOrUnload()` which looks like it will exclude all undesirable orders (when combined with the `OT_IMPLICIT` check).